modules/apps/portal/portal-messaging/src/main/java/com/liferay/portal/messaging/internal/DefaultDestinationFactory.java

```diff
@@ -80,34 +81,60 @@ protected void activate() {
 		_destinationPrototypes.put(
 			DestinationConfiguration.DESTINATION_TYPE_SYNCHRONOUS,
 			new SynchronousDestinationPrototype());
-	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void addDestinationPrototype(
-		DestinationPrototype destinationPrototype,
-		Map<String, Object> properties) {
+		String key = "destination.type";
 
-		_destinationPrototypes.put(
-			MapUtil.getString(properties, "destination.type"),
-			destinationPrototype);
+		_serviceTracker = ServiceTrackerFactory.open(
+			bundleContext, DestinationPrototype.class,
+			new ServiceTrackerCustomizer
+				<DestinationPrototype, DestinationPrototype>() {
+
+				@Override
+				public DestinationPrototype addingService(
+					ServiceReference<DestinationPrototype> serviceReference) {
+
+					DestinationPrototype destinationPrototype =
+						bundleContext.getService(serviceReference);
+
+					_destinationPrototypes.put(
+						MapUtil.getString(
```

Comment:
@W2jx Please don't create a temp singleton map for this. You can use `GetterUtil.getString(value)` instead.

Reply:
@W2jx And why this can't be a `ServiceTrackerMap`?

Reply:
Hi @dantewang , MapUtil.getString and GetterUtil.getString have different logic, so I didn't replace it with GetterUtil. Can I replace it?

Reply:
Because I keep the logic of removedService, which requires constructing a new ServiceTrackerCustomizer, and I think that since ServiceTrackerCustomizer has been constructed, using servicetracker directly can reduce additional encapsulation.